File: python/backtracking/nqueues.py
from typing import List


# N皇后

class Solution:

    def __init__(self):
        self.shu = set()
        self.pie = set()
        self.na = set()

    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []
        self.helper(n, 0, [], res)
        return res

    def helper(self, n, row, data, res):
        if row == n:
            temp = []
            board = [['.'] * n for _ in range(n)]
            for x, y in data:
                board[x][y] = 'Q'
                temp.append(''.join(board[x]))
            res.append(temp)
            return
        
        for col in range(n):
            if col in self.shu or col + row in self.pie or col - row in self.na:
                continue
            self.shu.add(col)
            self.pie.add(col + row)
            self.na.add(col - row)

            data.append((row, col))
            self.helper(n, row + 1, data, res)
            data.pop()
            
            self.shu.remove(col)
            self.pie.remove(col + row)
            self.na.remove(col - row)
            

File: go/backtracking/nqueues.go
package main

//N皇后问题
func solveNQueens(n int) [][]string {
	var res [][]string
	data := make([]int, n)
	shu, pie, na := make(map[int]bool), make(map[int]bool), make(map[int]bool)
	nQueuesHelper(n, 0, data, &res, shu, pie, na)
	return res
}

func nQueuesHelper(n, row int, data []int, res *[][]string, shu, pie, na map[int]bool) {
	if row == n {
		var temp []string
		for _, v := range data {
			var s []rune
			for i := 0; i < n; i++ {
				if i == v {
					s = append(s, 'Q')
				} else {
					s = append(s, '.')
				}
			}
			temp = append(temp, string(s))
		}
		*res = append(*res, temp)
		return
	}

	for col := 0; col < n; col++ {
		if !shu[col] && !pie[row+col] && !na[row-col] {
			shu[col], pie[row+col], na[row-col] = true, true, true
			data[row] = col
			nQueuesHelper(n, row+1, data, res, shu, pie, na)
			shu[col], pie[row+col], na[row-col] = false, false, false
		}
	}
}
